lib/shared/option-utils.js

```diff
@@ -0,0 +1,47 @@
+/**
+ * @fileoverview Utilities to operate on arrays.
+ * @author Josh Goldberg
+ */
+
+"use strict";
+
+/**
+ * Determines whether any of input's properties are different
+ * from values that already exist in original.
+ * @template T
+ * @param {Partial<T>} input New value.
+ * @param {T} original Original value.
+ * @returns {boolean} Whether input includes an explicit difference.
+ */
+function containsDifferentProperty(input, original) {
+    if (input === original) {
+        return false;
+    }
+
+    if (Array.isArray(input)) {
+        return (
+            !Array.isArray(original) ||
+            input.length > original.length ||
+            input.some((value, i) =>
+                containsDifferentProperty(value, original[i]))
+        );
+    }
+
+    if (typeof input !== typeof original) {
+        return true;
+    }
+
+    if (typeof input === "object") {
+        return Object.entries(input).some(
+            ([key, value]) =>
+                !Object.hasOwn(original, key) ||
+                containsDifferentProperty(value, original[key])
+        );
+    }
+
+    return true;
+}
```

Comment:
Arrays and objects should be compared entirely. For example, this is a false positive:

```js
// eslint.config.js

export default [{
    linterOptions: {
        reportUnusedInlineConfigs: "warn"
    },
    rules: {
        "id-denylist": ["error", "foo", "bar"]
    }
}];

```

```js
// my-file.js

/* eslint id-denylist: ["error", "foo"] */

foo; // id-denylist error on `foo`

bar; // no id-denylist error on `bar` because of the inline config

```

```
3:1  warning  "id-denylist" is already configured to error with the same options
```